### PR TITLE
MoveFile deal with destination which is a directory

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
That used to crash the app/test, because contentEquals require both files to be
either null or actual files (and not directory)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/scopedstorage/MigrateUserData.kt [hunk 1]
```diff
@@ -39,7 +39,12 @@ class MigrateUserData {
      *
      * If a file named `filename` exists in [destination] and in [source] with different content, move `source/filename` to `source/conflict/filename`.
      */
-    class FileConflictException(val source: DiskFile, val destination: DiskFile) : RuntimeException()
+    class FileConflictException(val source: DiskFile, val destination: DiskFile) : RuntimeException("File $source can not be copied to $destination, it exists and differs.")
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/scopedstorage/MigrateUserData.kt:42]:
```suggestion
    class FileConflictException(val source: DiskFile, val destination: DiskFile) : RuntimeException("File $source can not be copied to $destination, destination exists and differs.")
```

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/scopedstorage/MigrateUserData.kt [hunk 2]
```diff
@@ -39,7 +39,12 @@ class MigrateUserData {
...
+
+    /**
+     * If [destination] is a folder. In this case, move `source/filename` to `source/conflict/filename`.
+     */
+    class FileDirectoryConflictException(val source: DiskFile, val destination: Directory) : RuntimeException("File $source can not be copied to $destination, as it is a directory.")
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/scopedstorage/MigrateUserData.kt:47]:
```suggestion
    class FileDirectoryConflictException(val source: DiskFile, val destination: Directory) : RuntimeException("File $source can not be copied to $destination, as destination is a directory.")
```

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/scopedstorage/MoveFile.kt [hunk 3]
```diff
@@ -42,6 +42,7 @@ internal data class MoveFile(val sourceFile: DiskFile, val destinationFile: File
 
         if (destinationExists && destinationFile.isDirectory) {
             context.reportError(this, MigrateUserData.FileDirectoryConflictException(sourceFile, Directory.createInstanceUnsafe(destinationFile)))
+            return operationCompleted()
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/scopedstorage/MoveFile.kt:45]: seems this should be part of the first commit?

**Review by david-allison — APPROVED** [review]
Looks good to me. It seems like a very unlikely event, but it's not on the critical path, so the associated slowdown isn't a problem at all.

**Review by mikehardy — APPROVED** [review]
LGTM!

**Comment by Arthur-Milchior** [comment]
Let me assure you that this unlikely event occurred in test. Admittedly, it occurred only because I missed that the semantic of destination is different in MoveFileOrDirectory in one case and in MoveFile, MoveDirectory in the other case.
But really, this would have saved me an hour of debugging

**Comment by mikehardy** [comment]
#5304

**Pull request merged**